Add a dockerfile for running a set of Synapse worker processes (#9162)
matrix-org/synapse: docker/README.md

* `SYNAPSE_SERVER_NAME` (mandatory): the server public hostname.
* `SYNAPSE_REPORT_STATS` (mandatory, `yes` or `no`): whether to enable
  anonymous statistics reporting.

> * `SYNAPSE_HTTP_PORT`: the port Synapse should start listen for http traffic on.

Reviewer: should document the default here?

Reviewer: ```suggestion
* `SYNAPSE_HTTP_PORT`: the port Synapse should listen on for http traffic.
```

    matrixdotorg/synapse:latest
```

> assuming 8008 is the port Synapse is listening for traffic on.

Reviewer: ```suggestion
assuming 8008 is the port Synapse is configured to listen on for http traffic.
```

## Using jemalloc

Jemalloc is embedded in the image and will be used instead of the default allocator.

> You can read about jemalloc by reading the Synapse [README](../README.md)

Reviewer: ```suggestion
You can read about jemalloc by reading the Synapse [README](../README.md).
```

> ### Running all worker processes in a single container

Reviewer 2: ```suggestion
### Running all worker processes in a single container for testing
```

Just to make it *super* clear

> ### Running all worker processes in a single container for testing

Reviewer: I feel like this should be a separate file. This file is copied to https://hub.docker.com/r/matrixdotorg/synapse, and it's not really relevant to the main `matrixdotorg/synapse` image. (indeed, it would be rather confusing for people reading it at https://hub.docker.com/r/matrixdotorg/synapse)

Author: Oh indeed, I completely forgot that it ends up on Docker Hub. Good call.

Author: Although I'm struggling to come up with a good place to put it... Perhaps README-testing.md? Or maybe it should go in `docs/dev/` or CONTRIBUTING instead?

Reviewer: I'd go with a `README-testing.md` in the docker dir, since it's mostly describing how the docker image works. We can always add references to it from elsewhere (eg it would be good to stick a comment in the the dockerfile itself).

Author: I've now done this.

Anyone looking to run Synapse workers with docker in production should use one container
per Synapse process, specifying a different worker config per container.
Running all Synapse processes is possible, but is currently only used for testing 
purposes. [Dockerfile-workers](Dockerfile-workers) will produce an image bundling
all necessary components together for a workerised homeserver instance.

This includes any desired Synapse worker processes, a nginx to route traffic accordingly,
a redis for worker communication and a supervisord instance to start up and monitor all
processes. You will need to provide your own postgres container to connect to. TLS is not
handled by the container. A external reverse proxy should be placed in front the exposed
HTTP port for the purposes of TLS termination.

Note that the worker Dockerfile is based off the main `[Dockerfile](Dockerfile)`. If you
would like to build from the current checkout, first build the main Synapse docker image
using the instructions in [Building the image](#building-the-image), then afterwards build
the worker image:

```

> docker build -t matrixdotorg/synapse:workers -f docker/Dockerfile-workers .

Reviewer: given it's built from a different Dockerfile, I think it should have a different image name, rather than just being a tag on the regular image:

```suggestion
docker build -t matrixdotorg/synapse-workers -f docker/Dockerfile-workers .
```

```

To start a container, use the following:

```
docker run -d --name synapse \
    --mount type=volume,src=synapse-data,dst=/data \
    -p 8008:8008 \
    -e SYNAPSE_SERVER_NAME=my.matrix.host \
    -e SYNAPSE_REPORT_STATS=yes \
    -e SYNAPSE_WORKERS=synchrotron,media_repository,user_dir \
    matrixdotorg/synapse:workers
```

The `SYNAPSE_WORKERS` environment variable is a comma-separated list of workers to use
when running the container. All possible worker names are defined by the keys of the
`WORKERS_CONFIG` variable in [this script](configure_workers_and_start.py), which the
Dockerfile makes use of to generate appropriate worker, nginx and supervisord config files.

Sharding is supported for a subset of workers, in line with the [worker documentation]
(../docs/workers.md). To run multiple instances of a given worker type, simply specify 
the type multiple times in `SYNAPSE_WORKERS`
(e.g `SYNAPSE_WORKERS=event_creator,event_creator...`).

> Otherwise, `SYNAPSE_WORKERS` can either be left empty or unset to spawn no workers (
> leaving only the main process). The container is configured to use redis-based worker

Reviewer: ```suggestion
Otherwise, `SYNAPSE_WORKERS` can either be left empty or unset to spawn no workers 
(leaving only the main process). The container is configured to use redis-based worker
```
